File: Striver Important Interview Questions/Day 9 - Recursion/combination-sum.cpp

// https://leetcode.com/problems/combination-sum/description/

class Solution {
public:

    void solve(vector<int>&candidates, int target, vector<vector<int>>&ans, vector<int>&smallAns, int i){

        if(target == 0){
            ans.push_back(smallAns);
            return;
        }
        if(i >= candidates.size()) return;

        if(candidates[i] <= target){
            smallAns.push_back(candidates[i]);
            solve(candidates, target - candidates[i], ans, smallAns, i);
            smallAns.pop_back();
        }
        solve(candidates, target, ans, smallAns, i+1);
    }


    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n = candidates.size();
        vector<vector<int>> ans;
        if(n == 0) return ans;

        vector<int> smallAns;
        solve(candidates, target, ans, smallAns, 0);
        return ans;
    }
};

File: Neetcode-150-js/Backtracking/combination-sum.js

// https://leetcode.com/problems/combination-sum/

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let ans = [];
    let len = candidates.length;
    if(len === 0) return ans;

    let subset = [];

    var solve = (target, n) => {

        if(target === 0){
            ans = [...ans, [...subset]];
            return;
        }
        
        if(n === 0 || target < 0) return;

        if(candidates[n-1] <= target){
            subset.push(candidates[n-1]);
            solve(target - candidates[n-1], n);
            subset.pop();
        }
        solve(target, n-1);
    }

    solve(target, len);
    return ans;
};